File: Difficulty/Hard/[154] Find Minimum in Rotated Sorted Array II/solution.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        left, right, res = 0, len(nums) - 1, 2**31 - 1
        while left < right:
            if nums[left] < nums[right]:
                return min([res, nums[left], nums[0]])
            elif nums[left] == nums[right]:
                left += 1
                right -= 1
            else:
                mid = (left + right) // 2
                if nums[mid] < nums[left]:
                    right = mid
                else:
                    left = mid + 1
        return min([res, nums[left], nums[0]])

File: Difficulty/Hard/[154] Find Minimum in Rotated Sorted Array II/solution.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {
        int left = 0, right = nums.size() - 1;
        int result = INT_MAX;
        while (left < right) {
            if (nums[left] < nums[right])
                return min(min(result, nums[left]), nums[0]);
            else if (nums[left] == nums[right]) {
                left++;
                right--;
            }else {
                int mid = (left + right) / 2;
                if (nums[mid] < nums[left])
                    right = mid;
                else
                    left = mid + 1;
            }
        }
        return min(min(result, nums[left]), nums[0]);
    }
};